src/front/wgsl/lexer.rs

```diff
@@ -13,47 +13,328 @@ fn consume_any(input: &str, what: impl Fn(char) -> bool) -> (&str, &str) {
+                    true
+                }
+                _ => false,
+            },
+
+            NumberLexerState {
+                digit_state: NLDigitState::Exponent,
+                uint_suffix: false,
+                ..
+            } => match c {
+                '0'..='9' => {
+                    state.digit_state = NLDigitState::DigitAfterExponent;
+                    true
+                }
+                '-' | '+' => {
+                    state.digit_state = NLDigitState::SignAfterExponent;
+                    true
+                }
+                _ => false,
+            },
+
+            NumberLexerState {
+                digit_state: NLDigitState::SignAfterExponent | NLDigitState::DigitAfterExponent,
+                uint_suffix: false,
+                ..
+            } => match c {
+                '0'..='9' => {
+                    state.digit_state = NLDigitState::DigitAfterExponent;
+                    true
+                }
+                _ => false,
+            },
+
+            NumberLexerState {
+                uint_suffix: true, ..
+            } => false, // Scanning is done once we've reached a type suffix.
         }
     };
-    let pos = input.find(|c| !what(c)).unwrap_or_else(|| input.len());
-    let (value, rest) = input.split_at(pos);
-
-    let mut rest_iter = rest.chars();
-    let ty = rest_iter.next().unwrap_or(' ');
-    match ty {
-        'u' | 'i' | 'f' => {
-            let width_end = rest_iter
-                .position(|c| !('0'..='9').contains(&c))
-                .unwrap_or_else(|| rest.len() - 1);
-            let (width, rest) = rest[1..].split_at(width_end);
-            (Token::Number { value, ty, width }, rest)
-        }
-        // default to `i32` or `f32`
-        _ => (
+
+    let pos = working_substr
+        .find(|c| !what(c))
+        .unwrap_or_else(|| working_substr.len());
+    let (value, rest) = input.split_at(pos + minus_offset + hex_offset);
+
+    // NOTE: This code can use string slicing,
+    //       because number literals are exclusively ASCII.
+    //       This means all relevant characters fit into one byte
+    //       and using string slicing (which slices UTF-8 bytes) works for us.
+
+    if state.is_valid_number() {
+        (
+            Token::Number {
+                value: if state.uint_suffix {
+                    &value[..value.len() - 1]
+                } else {
+                    value
+                },
+                ty: if state.uint_suffix {
+                    'u'
+                } else if state.is_float() {
+                    'f'
+                } else {
+                    'i'
+                },
+                width: "",
+            },
+            rest,
+        )
+    } else {
+        // TODO: A syntax error can already be recognized here, possibly report it at this stage.
+
+        // Return known incorrect token for now.
+        (
             Token::Number {
```

Comment:
can we have a helper method for this? looks like it's getting repeated

Reply:
I could just remove the check for validity at that point in the code and return the Token, incorrect or not. The two branches are indeed the same.